File: electro-server/src/main/java/com/electro/entity/inventory/Docket.java
package com.electro.entity.inventory;

import com.electro.entity.BaseEntity;
import com.electro.entity.order.Order;
import com.fasterxml.jackson.annotation.JsonBackReference;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import java.util.HashSet;
import java.util.Set;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Accessors(chain = true)
@Entity
@Table(name = "docket")
public class Docket extends BaseEntity {
    @Column(name = "type", nullable = false)
    private Integer type;

    @Column(name = "code", nullable = false, unique = true)
    private String code;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reason_id", nullable = false)
    @JsonBackReference
    private DocketReason reason;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "warehouse_id", nullable = false)
    @JsonBackReference
    private Warehouse warehouse;

    @OneToMany(mappedBy = "docket", cascade = CascadeType.ALL)
    private Set<DocketVariant> docketVariants = new HashSet<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "purchase_order_id")
    @JsonBackReference
    private PurchaseOrder purchaseOrder;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    @JsonBackReference
    private Order order;

    @Column(name = "note")
    private String note;

    @Column(name = "status", nullable = false, columnDefinition = "TINYINT")
    private Integer status;
}

File: electro-client/src/models/Docket.ts
import BaseResponse from 'models/BaseResponse';
import { DocketReasonResponse } from 'models/DocketReason';
import { WarehouseResponse } from 'models/Warehouse';
import { DocketVariantRequest, DocketVariantResponse } from 'models/DocketVariant';

export interface DocketResponse extends BaseResponse {
  type: number;
  code: string;
  reason: DocketReasonResponse;
  warehouse: WarehouseResponse;
  docketVariants: DocketVariantResponse[];
  purchaseOrder: PurchaseOrderResponse | null;
  order: OrderResponse | null;
  note: string | null;
  status: number;
}

interface PurchaseOrderResponse extends BaseResponse {
  code: string;
  status: number;
}

interface OrderResponse extends BaseResponse {
  code: string;
  status: number;
}

export interface DocketRequest {
  type: number;
  code: string;
  reasonId: number;
  warehouseId: number;
  docketVariants: DocketVariantRequest[];
  purchaseOrderId: number | null;
  orderId: number | null;
  note: string | null;
  status: number;
}
